crates/unftp-sbe-gcs/src/lib.rs

```diff
@@ -134,30 +138,102 @@ impl CloudStorage {
-        CloudStorage {
+        Self {
             client,
             auth: auth.into(),
             uris: GcsUri::new(base_url.into(), bucket.into(), root),
+
+            cached_token: Default::default(),
         }
     }
 
-    // TODO: Cache the token. For `ServiceAccountKey`, the oauth client would already cache the token - we just need to move it to `CloudStorage`. For `WorkloadIdentity`, we can cache it in `CloudStorage`.
     #[tracing_attributes::instrument]
-    async fn get_token(&self) -> Result<String, Error> {
+    async fn get_token_value(&self) -> Result<String, Error> {
+        if let Some(token) = self.cached_token.get().await {
+            return Ok(token.value);
+        }
+
+        let token = self.fetch_token().await?;
+        self.cached_token.set(token.clone()).await;
+        Ok(token.value)
+    }
+
+    async fn fetch_token(&self) -> Result<Token, Error> {
         match &self.auth {
             AuthMethod::ServiceAccountKey(_) => {
                 let key = self.auth.to_service_account_key()?;
                 let auth = ServiceAccountAuthenticator::builder(key).hyper_client(self.client.clone()).build().await?;
 
                 auth.token(&["https://www.googleapis.com/auth/devstorage.read_write"])
-                    .map_ok(|t| t.as_str().to_string())
+                    .map_ok(|t| t.into())
                     .map_err(|e| Error::new(ErrorKind::PermanentFileNotAvailable, e))
                     .await
             }
-            AuthMethod::WorkloadIdentity(service) => workflow_identity::request_token(service.clone(), self.client.clone())
-                .await
-                .map(|t| t.access_token),
-            AuthMethod::None => Ok("unftp_test".to_string()),
+            AuthMethod::WorkloadIdentity(service) => workload_identity::request_token(service.clone(), self.client.clone()).await.map(|t| t.into()),
+            AuthMethod::None => Ok(Token {
+                value: "unftp_test".to_string(),
+                expires_at: None,
+            }),
+        }
+    }
+}
+
+#[derive(Default, Clone, Debug)]
+struct CachedToken {
+    inner: Arc<RwLock<Option<Token>>>,
+}
+
+impl CachedToken {
+    // get returns a token if it's available and not expired, and None otherwise.
+    async fn get(&self) -> Option<Token> {
+        let cache = self.inner.read().await;
+        cache.as_ref().map(|token| if token.active() { Some(token.clone()) } else { None }).flatten()
+    }
+
+    async fn set(&self, token: Token) {
+        let mut cache = self.inner.write().await;
+        *cache = Some(token);
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct Token {
+    value: String,
+    expires_at: Option<time::OffsetDateTime>,
+}
+
+impl Token {
+    /// active yields true when the token is present and has not expired. In all other cases, it
+    /// returns false.
+    fn active(&self) -> bool {
```

Comment:
```suggestion
    fn get(&self) -> Option<Token> 
```

I think doing this would make `CachedToken::get` simpler

Reply:
It would, but it would also make the interface of Token a bit more confusing. I could add another method to Token, `get_if_active` (name tbd), that returns `Option<Token>`, but I'm not sure whether that's worth it. WDYT?

Reply:
839a58c implements this

Reply:
Why not simply `get` then? If it returns an option it's already clear that it can return None, right?

Reply:
Because just a `get` method on a Token that returns an option of a token sounds weird to me. What does it mean when I have a token, call `get` on it, and it returns `None`? Naming it `get_if_active` expresses more clearly what happens and why.

Reply:
We should probably check what is idiomatic / customary in Rust code. Get to me is enough wit the clear return type. We can address this later in a refactoring

Reply:
Getting `None` means it's not able to get it. Not sure if the name must reveal why.

Reply:
But my opinion on this isn't that strong, I'm happy with the feature and want to see it merged.

Reply:
Here are some API naming guidelines: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter